pkg/retention/timeseries.go

```diff
@@ -0,0 +1,145 @@
+package retention
+
+import (
+	"sort"
+	"time"
+)
+
+// timeseries is a unique, sorted slice of Times representing
+// when snapshots were taken. It should be constructed
+// using make() followed by it's Push() method.
+type timeseries []time.Time
+
+// Implement sort.Interface:
+
+func (s timeseries) Len() int           { return len(s) }
+func (s timeseries) Less(i, j int) bool { return s[i].Before(s[j]) }
+func (s timeseries) Swap(i, j int)      { s[i], s[j] = s[j], s[i] }
+
+// Push inserts a one or more new snapshot Times in sorted order.
+func (sp *timeseries) Push(times ...time.Time) {
+	if len(times) == 0 {
+		return
+	}
+	sortedTimes := timeseries(times)
+	sortedTimes.squash()
+	s := *sp
+	if len(s) == 0 {
+		*sp = sortedTimes
+		return
+	}
+	s = append(s, sortedTimes...)
+	s.squash()
+	*sp = s
+}
+
+// Pop removes the time at an index are returns it.
+// It panics if the index is not present.
+func (s *timeseries) Pop(idx int) time.Time {
+	t := (*s)[idx]
+	if idx == 0 {
+		*s = (*s)[1:]
+		return t
+	}
+	if n := len(*s); idx == n-1 {
+		*s = (*s)[:n-1]
+		return t
+	}
+	*s = append((*s)[:idx], (*s)[idx+1:]...)
+	return t
+}
+
+// Discard removes the specified snapshot Times from the
+// timeseries if they are present.
+func (s *timeseries) Discard(ts ...time.Time) {
+	for _, t := range ts {
+		if i := s.Find(t); i != -1 {
+			_ = s.Pop(i)
+		}
+	}
+}
+
+// PopOldest pops the oldest snapshot off the slice
+// and returns it. If the slice is empty it returns
+// the zero Time.
+func (s *timeseries) PopOldest() time.Time {
+	if len(*s) == 0 {
+		return time.Time{}
+	}
+	return s.Pop(0)
+}
+
+// PeakOldest returns the oldest snapshot in the slice.
+// If the slice is empty it returns the zero Time.
+func (s timeseries) PeakOldest() time.Time {
```

Comment:
Should this function be named `Peek` instead of `Peak`?  (as in it lets you view without popping?)

Reply:
Yep, you are right.